[Step 4] 자동차 경주(우승자)

y2gcoder opened the pull request:
범준님 오늘 추웠는데 따뜻하게 보내셨는지요!

4단계 코드리뷰 부탁드리겠습니다!

번외로 범준님께서 3단계에서 던져주셨던 질문은 README.md에 적어놓았습니다! 

제가 이번 주에 게을러서 시간이 촉박했는지 바로 결정이 서지 않아 범준님께서 4단계 다른 요구사항에 대한 피드백을 주시는 동안 생각해보려고 합니다! 

항상 꼼꼼하게 봐주시고 생각할 수 있게 화두를 던져주셔서 감사합니다 :) 

양영근 올림

y2gcoder committed:
docs(racing-game): Define requirements for step4 in README.md and copy all files step3 to step4.racinggame

README.md에 step 4를 위한 요구사항 정리(기능, 프로그래밍, step3 리뷰어님 피드백)
분리 작업을 위한 step3 파일들을 step4.racinggame 패키지로 복사

y2gcoder committed:
feature(racing-game): Add name to Car and print roundResults with car's name

기능 요구사항 1 반영: 자동차 이름 부여
- Car 프로퍼티 추가: val name
- Car name 프로퍼티 유효성 검사 추가: 빈 문자열, 5글자 이하
- CarRoster에서 자동차 이름이 담긴 리스트 carNames를 파라미터로 받도록 변경
- CarRoster에서 자동차 이름 리스트로 자동차 목록을 생성하도록 책임 부여 및 메서드(createCars) 생성
- RacingGame: CarRoster로 Car 목록을 만들도록 변경
- InputView: inputCarNames, inputRound 함수 추가

y2gcoder committed:
feature(racing-game): Add getWinners to RacingGame and Add announceRaceWinners to RacingGameRaceResultView

기능 요구사항 2 반영: 우승자 발표
- RacingGame: getWinners() 추가
- RacingGameMain: RacingGame.getWinners() 호출
- RacingGameRaceResultView: 레이싱 게임 우승자들의 이름을 출력하는 announceRaceWinners() 추가

y2gcoder committed:
docs(racing-game): Make README.md's programmatic requirement 1,2 checked

README.md의 프로그래밍 요구사항 1, 2 체크

y2gcoder committed:
refactor(racing-game): Refactor functions to maintain an indent depth of 1

프로그래밍 요구사항 3 - Indent depth 를 1로 만들기 에 따라 코드 리팩터링
- RacingGame.moveAllCarPerRound()

y2gcoder committed:
refactor(racing-game): Refactor functions to ensure each is 15 lines or fewer

프로그래밍 요구사항 4 - 함수 라인 15줄 이하 에 따라 코드 리팩터링(함수가 하나의 일만 하도록)
- Car: init 블록과 move() 내 기능들을 별도의 메서드로 추출하여 리팩토링

y2gcoder committed:
docs(racing-game): Make README.md's programmatic requirement 5 checked

프로그래밍 요구사항 5 체크: 기능 목록 단위로 commit
리팩토링은 기능 단위가 아니라 생각해서 따로 커밋 진행함.
리뷰어님 피드백 필요

ddaaac commented on src/main/kotlin/week1/step4/racinggame/Car.kt:
```diff
@@ -0,0 +1,42 @@
+package week1.step4.racinggame
+
+class Car(val name: String, position: Int = DEFAULT_POSITION) {
```
저번 pr에서 눈치채지 못했는데 완전 다시 구현하는 식으로 미션을 진행하셨군요!

다음 단계?부터는 기존 클래스를 수정하는 쪽으로 진행해보시면 좋을 것 같아요.
깔끔한 코드를 짜고 싶다면 물론 처음부터 작성하는 게 편하겠지만, 저희는 대부분 시간을 기존 코드를 수정하는 데 시간을 많이 쓰다보니
교육을 받으면서도 어떻게 기존 코드를 잘 변경할 수 있을까? 또는 이런 코드를 짜야 변경이 쉽구나 / 어렵구나를 느껴볼 수 있기 위해서요!

y2gcoder replied:
넵 그렇게 진행하겠습니다! 감사합니다!

ddaaac commented on src/main/kotlin/week1/step4/racinggame/Car.kt:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    private fun validateRandomNumber(randomNumber: Int) {
```
물론 name, randomNumber에 대한 검증을 Car에서 할 수도 있겠지만 조금 더 책임을 나눠보는 것도 좋아보여요.
[value class](https://kotlinlang.org/docs/inline-classes.html)를 활용해볼까요?

ddaaac commented on src/main/kotlin/week1/step4/racinggame/RacingGame.kt:
```diff
@@ -0,0 +1,56 @@
+
+import kotlin.random.Random
+
+@JvmInline
+value class CarRoster(private val carNames: List<String>) {
+    init {
+        require(carNames.isNotEmpty()) {
+            "레이싱 게임을 하기 위해서는 차가 1대 이상이어야 합니다."
+        }
+    }
+
+    fun createCars(): List<Car> {
+        return carNames.map { Car(it) }
+    }
+}
+
+fun interface RandomNumberGenerator {
+    fun generate(): Int
+}
```
따로 파일 분리를 해도 좋을 것 같아요. 혹시 같은 파일에 두신 이유가 있을까요?

y2gcoder replied:
저난 번 피드백 때 참고하라고 보내주신 링크가 inline value class 였고, 뭔가 RacingGame 과 같이 사용할 것 같아 같은 곳에 모아두는 게 어떨까 해서 같은 파일에 담아두었습니다! CarRoster 는 밖으로 빼보겠습니다!

ddaaac commented on src/main/kotlin/week1/step4/racinggame/RacingGame.kt:
```diff
@@ -0,0 +1,56 @@
+
+class RacingGame(
+    carRoster: CarRoster,
+    private val randomNumberGenerator: RandomNumberGenerator = RandomNumberGenerator { Random.nextInt(0, 10) },
```
인터페이스를 파라미터로 받아서 테스트 가능한 구조를 만들어주셨군요 👍

ddaaac commented on src/main/kotlin/week1/step4/racinggame/RacingGameRaceResultView.kt:
```diff
@@ -0,0 +1,24 @@
+        }
+
+        private fun printRoundResult(roundResult: RoundResult) {
+            roundResult.result.forEach { (car, position) ->
```
data class의 [destructuring](https://kotlinlang.org/docs/destructuring-declarations.html)을 잘 활용해주셨군요 👍

ddaaac commented on src/test/kotlin/week1/step4/racinggame/RacingGameTest.kt:
```diff
@@ -0,0 +1,78 @@
+
+        val roundResults = sut.race(round)
+
+        roundResults.size shouldBe round
+        roundResults.forEachIndexed { index, roundResult ->
+            val expectedPosition = index + 1
+            roundResult.result.forEach { (car, roundPosition) ->
+                car.name shouldBeIn carNames
+                roundPosition shouldBe expectedPosition
+            }
+        }
+    }
```
전체적으로 tc가 꼼꼼해서 좋네요 👍 

이 부분의 테스트가 읽기 쉬운지 고민해봐도 좋을 것 같아요. 테스트가 읽기 좋다 = 메서드의 시그니처가 명확하고 하는 일이 분명하다 라고 생각할 수도 있을텐데요, Map보다는 라운드 하나씩의 결과를 반환하는 게 테스트 코드가 더 깔끔하고 이해하기 쉬운 코드일지도 모르겠다는 생각이 드네요. 영근님의 생각은 어떠신가요?

ddaaac commented on src/main/kotlin/week1/step4/racinggame/RacingGame.kt:
```diff
@@ -0,0 +1,56 @@
+
+    private fun getRoundResult() = RoundResult(cars.associateWith { it.position })
+
+    fun getWinners(): List<Car> {
+        val winnerPosition = cars.maxOf { it.position }
+        return cars.filter { it.position == winnerPosition }
+    }
```
지금 우승자를 구하는 로직도 좋지만 객체지향을 학습하는 교육과정이니만큼 조금 더 극단적으로 구현해볼 수 있지 않을까요?
Car 내부 프로퍼티인 position을 Cars에서 모르게 구현해보면 좋을 것 같아요.
예를 들면 같은 position인지 검사하는 로직은 car.isDraw(other), 다른 차를 이기는지 검사하는 로직은 car.isWin(other)과 같이 구현할 수 있겠죠.

ddaaac reviewed (changes requested):
안녕하세요 영근님 👋 

늦어서 죄송합니다 ㅜㅜ

전반적으로 너무 잘 작성해주셨는데요, 몇 가지 코멘트 남겼으니 확인해주세요~

y2gcoder committed:
feature(racing-game): Convert Java StringTest to Kotlin StringTest

1주차 비대면 강의 내용 중 Java로 잘못 작성한 StringTest 를 Kotlin 테스트로 변경

y2gcoder committed:
feature(racing-game): CarRoster 를 별개의 파일로 분리

리뷰어님 피드백 반영.

y2gcoder committed:
feature(racing-game): Car의 name 에 대한 value class CarName 생성 및 반영

리뷰어님 피드백 반영.
CarName 은 자동차 이름에 대한 유효성 검증, 자동차 이름 반환 등 자동차 이름과 관련된 역할을 담당하고 있음.

y2gcoder committed:
feature(racing-game): Car의 move()의 파라미터인 randomNumber 를 value class 로 래핑(RandomNumber)

리뷰어님 피드백 반영.
RandomNumber 가 입력할 숫자의 범위에 대한 유효성 검사를 담당(0..9)

y2gcoder committed:
feature(racing-game): RacingGame 의 레이스 방식을 한 라운드씩 실행하는 걸로 변경

리뷰어님 피드백 반영.
1라운드씩 실행하는 함수를 만들고, 각 라운드 실행 후 cars를 받아 라운드 결과를 표시하도록 변경

y2gcoder committed:
refactor(racing-game): RacingGame 의 getWinners 리팩토링하여 car의 position 은닉

리뷰어님 피드백 반영.
RacingGame의 getWinners() 를 구할 때 car.position 을 드러내지 않고 car.isWin(), car.isDraw()를 이용해 승자 목록 획득하도록 리팩토링

ddaaac commented on src/main/kotlin/week1/step4/racinggame/Car.kt:
```diff
@@ -0,0 +1,34 @@
+package week1.step4.racinggame
+
+class Car(name: String, position: Int = DEFAULT_POSITION) {
+    val name = CarName(name)
+    var position: Int = position
+        private set
+
+    fun move(randomNumber: RandomNumber) {
+        if (canMove(randomNumber.value)) {
+            position++
+        }
+    }
+
+    fun getName(): String {
+        return name.value
```
value class를 잘 활용해주셨군요 👍 

저는 개인적으로 값 객체를 사용할 때, 값 객체를 최대한 빨리 생성하고 최대한 늦게 그 값을 꺼내는 편입니다.
꼭 필요할 때 값을 꺼내서 사용해야하는 느낌이죠. 예를 들면 처음 값이 input view를 통해 들어올 때 바로 값 객체로 포장하고, 마지막에 View에서 사용할 때가 돼서야 값을 꺼내도록요.

값 객체로 감싸는 의미가 이 값에 특정 의미를 부여하거나 검증을 하기 위해서인데 
이 값이 어플리케이션 중간에 생성되거나 꺼내지면, 객체로 포장한 이유가 덜해지지 않나 싶어요.
영근님의 생각은 어떠신가요?

y2gcoder replied:
래핑하는 의미를 다시 한 번 생각해보게
되네요!
확실하게 이해가 됩니다! 다음 단계에서 반영해보겠습니다!

ddaaac reviewed (approved):
안녕하세요 영근님 👋 

구현을 너무 잘 해주셔서 크게 남길 코멘트가 없네요.

다음 단계 진행해주세요!
